corporate/tests/test_stripe.py

```diff
@@ -5722,11 +5740,140 @@ def test_non_sponsorship_billing(self, *mocks: Mock) -> None:
+        )
+
+        self.assertFalse(Customer.objects.exists())
+        self.assertFalse(CustomerPlan.objects.exists())
+        self.assertFalse(LicenseLedger.objects.exists())
+
+        with time_machine.travel(self.now, tick=False):
+            stripe_customer = self.add_card_and_upgrade(
+                tier=CustomerPlan.TIER_SELF_HOSTED_BASIC, schedule="monthly"
+            )
+
+        customer = Customer.objects.get(stripe_customer_id=stripe_customer.id)
+        plan = CustomerPlan.objects.get(customer=customer)
+        LicenseLedger.objects.get(plan=plan)
+
+        with time_machine.travel(self.now + timedelta(days=1), tick=False):
+            response = self.client_get(
+                f"{self.billing_session.billing_base_url}/billing/", subdomain="selfhosting"
+            )
+        for substring in [
+            "Zulip Basic",
+            "Number of licenses",
+            f"{realm_user_count} (managed automatically)",
+            "February 2, 2012",
+            "Your plan will automatically renew on",
+            f"${3.5 * realm_user_count - flat_discount // 100 * 1:,.2f}",
+            "Visa ending in 4242",
+            "Update card",
+        ]:
+            self.assert_in_response(substring, response)
+
+        # Verify that change in user count updates LicenseLedger.
+        audit_log_count = RemoteRealmAuditLog.objects.count()
+        self.assertEqual(LicenseLedger.objects.count(), 1)
+
+        with time_machine.travel(self.now + timedelta(days=2), tick=False):
+            for count in range(realm_user_count, min_licenses + 10):
+                do_create_user(
+                    f"email {count}",
+                    f"password {count}",
+                    hamlet.realm,
+                    "name",
+                    role=UserProfile.ROLE_MEMBER,
+                    acting_user=None,
+                )
+
+        with time_machine.travel(self.now + timedelta(days=3), tick=False):
+            send_server_data_to_push_bouncer(consider_usage_statistics=False)
+
+        self.assertEqual(
+            RemoteRealmAuditLog.objects.count(),
+            min_licenses + 10 - realm_user_count + audit_log_count,
+        )
+        latest_ledger = LicenseLedger.objects.last()
+        assert latest_ledger is not None
+        self.assertEqual(latest_ledger.licenses, min_licenses + 10)
+
+        with time_machine.travel(self.now + timedelta(days=1), tick=False):
+            response = self.client_get(
+                f"{self.billing_session.billing_base_url}/billing/", subdomain="selfhosting"
+            )
+
+        self.assertEqual(latest_ledger.licenses, 20)
+        for substring in [
+            "Zulip Basic",
+            "Number of licenses",
+            f"{latest_ledger.licenses} (managed automatically)",
+            "February 2, 2012",
+            "Your plan will automatically renew on",
+            f"${3.5 * latest_ledger.licenses - flat_discount // 100 * 1:,.2f}",
```

Comment:
I think for these it's often better to just write the answer into the test (with a comment for how it's computed); it's more robust in some ways to do that, since any change in the answer probably is worth explicitly noting.